File: packages/sdk/python/clouddevbox/apikeys.py
"""API Keys Service"""

from dataclasses import dataclass
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import DevBoxClient, AsyncDevBoxClient


@dataclass
class APIKey:
    """API key"""
    id: str
    name: str
    key_prefix: str
    scopes: list[str]
    is_active: bool
    created_at: str
    description: Optional[str] = None
    expires_at: Optional[str] = None
    last_used_at: Optional[str] = None
    last_used_ip: Optional[str] = None


class APIKeysService:
    """API Keys service for sync client"""

    def __init__(self, client: "DevBoxClient"):
        self._client = client

    def list(self, page: int = 1, page_size: int = 20) -> dict:
        """List API keys"""
        return self._client.request("GET", "/api-keys", params={"page": page, "page_size": page_size})

    def get(self, id: str) -> dict:
        """Get API key by ID"""
        return self._client.request("GET", f"/api-keys/{id}")

    def create(self, name: str, scopes: list[str], **kwargs) -> dict:
        """Create a new API key"""
        data = {"name": name, "scopes": scopes, **kwargs}
        return self._client.request("POST", "/api-keys", json=data)

    def update(self, id: str, **kwargs) -> dict:
        """Update an API key"""
        return self._client.request("PUT", f"/api-keys/{id}", json=kwargs)

    def delete(self, id: str) -> None:
        """Delete an API key"""
        self._client.request("DELETE", f"/api-keys/{id}")

    def revoke(self, id: str) -> None:
        """Revoke an API key"""
        self._client.request("POST", f"/api-keys/{id}/revoke")

    def get_available_scopes(self) -> list[str]:
        """Get available API key scopes"""
        return self._client.request("GET", "/api-keys/scopes")


class AsyncAPIKeysService:
    """API Keys service for async client"""

    def __init__(self, client: "AsyncDevBoxClient"):
        self._client = client

    async def list(self, page: int = 1, page_size: int = 20) -> dict:
        """List API keys"""
        return await self._client.request("GET", "/api-keys", params={"page": page, "page_size": page_size})

    async def get(self, id: str) -> dict:
        """Get API key by ID"""
        return await self._client.request("GET", f"/api-keys/{id}")

    async def create(self, name: str, scopes: list[str], **kwargs) -> dict:
        """Create a new API key"""
        data = {"name": name, "scopes": scopes, **kwargs}
        return await self._client.request("POST", "/api-keys", json=data)

    async def update(self, id: str, **kwargs) -> dict:
        """Update an API key"""
        return await self._client.request("PUT", f"/api-keys/{id}", json=kwargs)

    async def delete(self, id: str) -> None:
        """Delete an API key"""
        await self._client.request("DELETE", f"/api-keys/{id}")

    async def revoke(self, id: str) -> None:
        """Revoke an API key"""
        await self._client.request("POST", f"/api-keys/{id}/revoke")

    async def get_available_scopes(self) -> list[str]:
        """Get available API key scopes"""
        return await self._client.request("GET", "/api-keys/scopes")

File: packages/sdk/nodejs/src/apikeys.ts
/**
 * API Keys Service
 */

import type { DevBoxClient } from './client';
import type { ListOptions, PaginatedResponse } from './types';

export interface APIKey {
  id: string;
  name: string;
  description?: string;
  key_prefix: string;
  scopes: string[];
  expires_at?: string;
  last_used_at?: string;
  last_used_ip?: string;
  is_active: boolean;
  created_at: string;
}

export interface CreateAPIKeyRequest {
  name: string;
  description?: string;
  scopes: string[];
  expires_at?: string;
}

export interface CreateAPIKeyResponse {
  api_key: APIKey;
  plain_key: string;
}

export interface UpdateAPIKeyRequest {
  name?: string;
  description?: string;
  scopes?: string[];
}

export class APIKeysService {
  constructor(private client: DevBoxClient) {}

  async list(options?: ListOptions): Promise<PaginatedResponse<APIKey>> {
    const params = new URLSearchParams();
    if (options?.page) params.set('page', String(options.page));
    if (options?.pageSize) params.set('page_size', String(options.pageSize));
    
    const query = params.toString();
    return this.client.request('GET', `/api-keys${query ? `?${query}` : ''}`);
  }

  async get(id: string): Promise<APIKey> {
    return this.client.request('GET', `/api-keys/${id}`);
  }

  async create(request: CreateAPIKeyRequest): Promise<CreateAPIKeyResponse> {
    return this.client.request('POST', '/api-keys', request);
  }

  async update(id: string, request: UpdateAPIKeyRequest): Promise<APIKey> {
    return this.client.request('PUT', `/api-keys/${id}`, request);
  }

  async delete(id: string): Promise<void> {
    return this.client.request('DELETE', `/api-keys/${id}`);
  }

  async revoke(id: string): Promise<void> {
    return this.client.request('POST', `/api-keys/${id}/revoke`);
  }

  async getAvailableScopes(): Promise<string[]> {
    return this.client.request('GET', '/api-keys/scopes');
  }
}
